File: src/main/java/mediathek/gui/dialogEinstellungen/PanelEinstellungenColor.java
package mediathek.gui.dialogEinstellungen;

import mediathek.config.MVColor;
import mediathek.mainwindow.MediathekGui;
import mediathek.tool.MVC;
import mediathek.tool.cellrenderer.CellRendererColor;
import mediathek.tool.models.TModelColor;
import net.miginfocom.layout.AC;
import net.miginfocom.layout.CC;
import net.miginfocom.layout.LC;
import net.miginfocom.swing.MigLayout;

import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;

public class PanelEinstellungenColor extends JPanel {
    private final TModelColor lightColorTableModel = new TModelColor(false);
    private final TModelColor darkColorTableModel = new TModelColor(true);

    public PanelEinstellungenColor() {
        initComponents();
        init();
    }

    /**
     * Force update of the user interface.
     */
    public void updateGui() {
        try {
            MediathekGui.ui().setupAlternatingRowColors();
            SwingUtilities.updateComponentTreeUI(MediathekGui.ui());
            for (Frame f : Frame.getFrames()) {
                SwingUtilities.updateComponentTreeUI(f);
                for (Window w : f.getOwnedWindows()) {
                    SwingUtilities.updateComponentTreeUI(w);
                }
            }
        } catch (Exception ignored) {
        }
    }

    private void init() {
        configureTable(jTableLight, lightColorTableModel);
        configureTable(jTableDark, darkColorTableModel);
        jTabbedPane1.addChangeListener(_ -> updateResetButtonText());
        jButtonReset.addActionListener(_ -> resetSelectedThemeColors());
        updateResetButtonText();
    }

    private void configureTable(JTable table, TModelColor model) {
        table.addMouseListener(new BeobMausTabelle(table, model));
        table.setDefaultRenderer(MVC.class, new CellRendererColor());
        table.setModel(model);
    }

    private void getColor(MVC mvc, boolean darkMode) {
        var selectedColor = JColorChooser.showDialog(this,"Farbe auswählen", mvc.getColor(darkMode));
        if (selectedColor != null) {
            if (!selectedColor.equals(mvc.getColor(darkMode))) {
                mvc.setColor(darkMode, selectedColor);
                refreshColorTables();
                updateGui();
                MVColor.save();
            }
        }
    }

    private void refreshColorTables() {
        lightColorTableModel.fireTableDataChanged();
        darkColorTableModel.fireTableDataChanged();
    }

    private void resetSelectedThemeColors() {
        boolean darkMode = jTabbedPane1.getSelectedIndex() == 1;
        for (MVC mvc : MVColor.getColors()) {
            mvc.reset(darkMode);
        }
        refreshColorTables();
        updateGui();
        MVColor.save();
    }

    private void updateResetButtonText() {
        jButtonReset.setText(jTabbedPane1.getSelectedIndex() == 1
                ? "Dunkle Farben zurücksetzen"
                : "Helle Farben zurücksetzen");
    }

    public class BeobMausTabelle extends MouseAdapter {
        private final JTable table;
        private final TModelColor model;

        public BeobMausTabelle(JTable table, TModelColor model) {
            this.table = table;
            this.model = model;
        }

        @Override
        public void mouseClicked(MouseEvent arg0) {
            if (arg0.getButton() == MouseEvent.BUTTON1) {
                if (arg0.getClickCount() == 1) {
                    Point p = arg0.getPoint();
                    int row = table.rowAtPoint(p);
                    int column = table.columnAtPoint(p);
                    if (row >= 0) {
                        MVC mvc = model.getEntry(table.convertRowIndexToModel(row));
                        if (table.convertColumnIndexToModel(column) == MVColor.MVC_COLOR) {
                            getColor(mvc, model.isDarkMode());
                        }
                    }
                }
            }
        }

    }


    // <editor-fold defaultstate="collapsed" desc="Generated Code">//GEN-BEGIN:initComponents
    // Generated using JFormDesigner non-commercial license
    private void initComponents() {
        jTabbedPane1 = new JTabbedPane();
        var panelLight = new JPanel();
        var scrollPaneLight = new JScrollPane();
        jTableLight = new JTable();
        var panelDark = new JPanel();
        var scrollPaneDark = new JScrollPane();
        jTableDark = new JTable();
        var hSpacer1 = new JPanel(null);
        jButtonReset = new JButton();

        //======== this ========
        setLayout(new MigLayout(
            new LC().insets("5").hideMode(3).gridGap("5", "5"), //NON-NLS
            // columns
            new AC()
                .fill().gap()
                .grow().fill().gap()
                .fill(),
            // rows
            new AC()
                .grow().fill().gap()
                .fill()));

        //======== jTabbedPane1 ========
        {
            //======== panelLight ========
            {
                panelLight.setLayout(new MigLayout(
                    new LC().insets("0").hideMode(3),
                    // columns
                    new AC()
                        .grow().fill(),
                    // rows
                    new AC()
                        .grow().fill()));

                //======== scrollPaneLight ========
                {
                    scrollPaneLight.setViewportView(jTableLight);
                }
                panelLight.add(scrollPaneLight, new CC().cell(0, 0));
            }
            jTabbedPane1.addTab("Hell", panelLight);

            //======== panelDark ========
            {
                panelDark.setLayout(new MigLayout(
                    new LC().insets("0").hideMode(3),
                    // columns
                    new AC()
                        .grow().fill(),
                    // rows
                    new AC()
                        .grow().fill()));

                //======== scrollPaneDark ========
                {
                    scrollPaneDark.setViewportView(jTableDark);
                }
                panelDark.add(scrollPaneDark, new CC().cell(0, 0));
            }
            jTabbedPane1.addTab("Dunkel", panelDark);
        }
        add(jTabbedPane1, new CC().cell(1, 0, 2, 1));
        add(hSpacer1, new CC().cell(1, 1));

        //---- jButtonReset ----
        jButtonReset.setText("Helle Farben zur\u00fccksetzen"); //NON-NLS
        add(jButtonReset, new CC().cell(2, 1));
    }// </editor-fold>//GEN-END:initComponents

    // Variables declaration - do not modify//GEN-BEGIN:variables
    // Generated using JFormDesigner non-commercial license
    private JTabbedPane jTabbedPane1;
    private JTable jTableLight;
    private JTable jTableDark;
    private JButton jButtonReset;
    // End of variables declaration//GEN-END:variables
}

File: src/main/kotlin/mediathek/gui/dialogEinstellungen/PanelEinstellungenColor.kt
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package mediathek.gui.dialogEinstellungen

import mediathek.config.MVColor
import mediathek.mainwindow.SettingsDialogHost
import mediathek.tool.MVC
import mediathek.tool.cellrenderer.CellRendererColor
import mediathek.tool.models.TModelColor
import java.awt.Frame
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import javax.swing.JColorChooser
import javax.swing.JTable
import javax.swing.SwingUtilities

class PanelEinstellungenColor(
    private val host: SettingsDialogHost,
) : PanelEinstellungenColorBase() {
    private val lightColorTableModel = TModelColor(isDarkMode = false)
    private val darkColorTableModel = TModelColor(isDarkMode = true)
    private val colorThemes = listOf(
        ColorTheme(
            tabIndex = 0,
            resetButtonText = "Helle Farben zurücksetzen",
            model = lightColorTableModel,
        ),
        ColorTheme(
            tabIndex = 1,
            resetButtonText = "Dunkle Farben zurücksetzen",
            model = darkColorTableModel,
        ),
    )

    init {
        configureTable(jTableLight, lightColorTableModel)
        configureTable(jTableDark, darkColorTableModel)
        jTabbedPane1.addChangeListener { updateResetButtonText() }
        jButtonReset.addActionListener { resetSelectedThemeColors() }
        updateResetButtonText()
    }

    /**
     * Force update of the user interface.
     */
    private fun updateGui() {
        try {
            host.setupAlternatingRowColors()
            SwingUtilities.updateComponentTreeUI(host.ownerFrame())
            for (frame in Frame.getFrames()) {
                SwingUtilities.updateComponentTreeUI(frame)
                for (window in frame.ownedWindows) {
                    SwingUtilities.updateComponentTreeUI(window)
                }
            }
        } catch (_: Exception) {
        }
    }

    private fun configureTable(table: JTable, model: TModelColor) {
        table.addMouseListener(ColorTableMouseListener(table, model))
        table.model = model
        table.columnModel
            .getColumn(MVColor.MVC_COLOR)
            .cellRenderer = CellRendererColor { model.isDarkMode }
    }

    private fun chooseColor(mvc: MVC, darkMode: Boolean) {
        val currentColor = mvc.getColor(darkMode)
        val selectedColor = JColorChooser.showDialog(this, "Farbe auswählen", currentColor)
        if (selectedColor != null && selectedColor != currentColor) {
            mvc.setColor(darkMode, selectedColor)
            saveAndRefreshColors()
        }
    }

    private fun saveAndRefreshColors() {
        lightColorTableModel.fireTableDataChanged()
        darkColorTableModel.fireTableDataChanged()
        updateGui()
        MVColor.save()
    }

    private fun resetSelectedThemeColors() {
        val darkMode = selectedTheme.model.isDarkMode
        for (mvc in MVColor.getColors()) {
            mvc.reset(darkMode)
        }
        saveAndRefreshColors()
    }

    private fun updateResetButtonText() {
        jButtonReset.text = selectedTheme.resetButtonText
    }

    private val selectedTheme: ColorTheme
        get() = colorThemes.firstOrNull { it.tabIndex == jTabbedPane1.selectedIndex } ?: colorThemes.first()

    private inner class ColorTableMouseListener(
        private val table: JTable,
        private val model: TModelColor,
    ) : MouseAdapter() {
        override fun mouseClicked(event: MouseEvent) {
            if (event.button != MouseEvent.BUTTON1 || event.clickCount != 1) {
                return
            }

            val row = table.rowAtPoint(event.point)
            val column = table.columnAtPoint(event.point)
            if (row < 0 || column < 0) {
                return
            }

            val mvc = model.getEntry(table.convertRowIndexToModel(row))
            if (table.convertColumnIndexToModel(column) == MVColor.MVC_COLOR) {
                chooseColor(mvc, model.isDarkMode)
            }
        }
    }

    private data class ColorTheme(
        val tabIndex: Int,
        val resetButtonText: String,
        val model: TModelColor,
    )
}
